csrc/mmdeploy/net/rknn/rknn_net.cpp

```diff
@@ -0,0 +1,211 @@
+// Copyright (c) OpenMMLab. All rights reserved.
+#include "rknn_net.h"
+
+#include <stdio.h>
+
+#include <fstream>
+
+#include "mmdeploy/core/logger.h"
+#include "mmdeploy/core/model.h"
+#include "mmdeploy/core/utils/filesystem.h"
+#include "mmdeploy/core/utils/formatter.h"
+
+namespace mmdeploy::framework {
+
+RKNNNet::~RKNNNet() {}
+
+void RKNNNet::dump_tensor_attr(rknn_tensor_attr* attr) {
+  printf(
+      "  index=%d, name=%s, n_dims=%d, dims=[%d, %d, %d, %d], n_elems=%d, size=%d, fmt=%s, "
+      "type=%s, qnt_type=%s, "
+      "zp=%d, scale=%f\n",
+      attr->index, attr->name, attr->n_dims, attr->dims[0], attr->dims[1], attr->dims[2],
+      attr->dims[3], attr->n_elems, attr->size, get_format_string(attr->fmt),
+      get_type_string(attr->type), get_qnt_type_string(attr->qnt_type), attr->zp, attr->scale);
+}
+
+unsigned char* RKNNNet::load_model(const char* filename, int* model_size) {
+  FILE* fp = fopen(filename, "rb");
+  if (fp == nullptr) {
+    printf("fopen %s fail!\n", filename);
+    return NULL;
+  }
+  fseek(fp, 0, SEEK_END);
+  int model_len = ftell(fp);
+  unsigned char* model = (unsigned char*)malloc(model_len);
+  fseek(fp, 0, SEEK_SET);
+  if (model_len != fread(model, 1, model_len, fp)) {
+    printf("fread %s fail!\n", filename);
+    free(model);
+    return NULL;
+  }
+  *model_size = model_len;
+  if (fp) {
+    fclose(fp);
+  }
+  return model;
+}
+
+Result<void> RKNNNet::Init(const Value& args) {
+  auto& context = args["context"];
+  device_ = context["device"].get<Device>();
+  stream_ = context["stream"].get<Stream>();
+  if (!device_.is_host()) {
+    return Status(eNotSupported);
+  }
+
+  auto name = args["name"].get<std::string>();
+  auto model = context["model"].get<Model>();
+  OUTCOME_TRY(auto config, model.GetModelConfig(name));
+
+  std::string content;
+  OUTCOME_TRY(content, model.ReadFile(config.net));
+  char* model_ptr = const_cast<char*>(content.data());
+  int ret = rknn_init(&ctx_, model_ptr, content.size(), 0, NULL);
+  if (ret < 0) {
+    MMDEPLOY_ERROR("Load .rknn failed: {}", config.net);
+    return Status(eInvalidArgument);
+  }
+
+  // Get Model Input Output Info
+  rknn_input_output_num io_num;
+  ret = rknn_query(ctx_, RKNN_QUERY_IN_OUT_NUM, &io_num, sizeof(io_num));
+  if (ret != RKNN_SUCC) {
+    MMDEPLOY_ERROR("rknn_query fail! ret= {}", ret);
+    return Status(eFail);
+  }
+  printf("model input num: %d, output num: %d\n", io_num.n_input, io_num.n_output);
+
+  printf("input tensors:\n");
+  rknn_tensor_attr input_attrs[io_num.n_input];
+  memset(input_attrs, 0, sizeof(input_attrs));
+  for (int i = 0; i < io_num.n_input; i++) {
+    input_attrs[i].index = i;
+    ret = rknn_query(ctx_, RKNN_QUERY_INPUT_ATTR, &(input_attrs[i]), sizeof(rknn_tensor_attr));
```

Comment:
May add "RKNN_QUERY_INPUT_ATTR" in the error log as well so that users can know which query failed.